build & ship assets as part of deployment

```
[staging.openoversight.lucyparsonslabs.com] out: Unpacking objects:  14% (1/7)   
[staging.openoversight.lucyparsonslabs.com] out: Unpacking objects:  28% (2/7)   
[staging.openoversight.lucyparsonslabs.com] out: Unpacking objects:  42% (3/7)   
[staging.openoversight.lucyparsonslabs.com] out: Unpacking objects:  57% (4/7)   
[staging.openoversight.lucyparsonslabs.com] out: Unpacking objects:  71% (5/7)   
[staging.openoversight.lucyparsonslabs.com] out: Unpacking objects:  85% (6/7)   
[staging.openoversight.lucyparsonslabs.com] out: Unpacking objects: 100% (7/7)   
[staging.openoversight.lucyparsonslabs.com] out: Unpacking objects: 100% (7/7), done.
[staging.openoversight.lucyparsonslabs.com] out: From https://github.com/lucyparsons/OpenOversight
[staging.openoversight.lucyparsonslabs.com] out:    90e5d86..74dc6b0  develop    -> origin/develop
[staging.openoversight.lucyparsonslabs.com] out: On branch develop
[staging.openoversight.lucyparsonslabs.com] out: Your branch is behind 'origin/develop' by 40 commits, and can be fast-forwarded.
[staging.openoversight.lucyparsonslabs.com] out:   (use "git pull" to update your local branch)
[staging.openoversight.lucyparsonslabs.com] out: nothing to commit, working directory clean
[staging.openoversight.lucyparsonslabs.com] out: 

[staging.openoversight.lucyparsonslabs.com] run: su nginx -c "make cleanassets assets"
[staging.openoversight.lucyparsonslabs.com] out: make: *** No rule to make target 'cleanassets'.  Stop.
[staging.openoversight.lucyparsonslabs.com] out: 
```
The current deployment process assumes the remote web servers have docker-compose installed to allow them to build the web assets.
I'd love to have them build locally as part of the deployment, and be scp'd up, since that saves us from running a dockerd or installing yarn/etc on the servers.

While we're doing this, we should probably rewrite the fabfile to be fabric 2-compatible, since fabric 1's support of python3 is not great. (I have to keep a python2 venv hanging around because my OS uses 3.7 by default, which fabric 1.x doesn't seem to support.)

--- a/fabfile.py
+++ b/fabfile.py
@@ -54,7 +54,7 @@ def deploy():
                 run('su %s -c "PATH=%s/bin:$PATH pip install -r requirements.txt"' % (env.unprivileged_user, env.venv_dir))
                 run('su %s -c "mkdir --parents %s/OpenOversight/app/static/dist"' % (env.unprivileged_user, env.code_dir))
                 put(local_path=os.path.join('OpenOversight', 'app', 'static', 'dist'),
-                    remote_path=os.path.join(env.code_dir, 'OpenOversight', 'app', 'static', 'dist')
+                    remote_path=os.path.join(env.code_dir, 'OpenOversight', 'app', 'static')
                     )
                 run('sudo systemctl restart openoversight')
 

--- a/fabfile.py
+++ b/fabfile.py
@@ -1,5 +1,5 @@
 from __future__ import with_statement
-from fabric.api import env, run, cd, get
+from fabric.api import env, run, cd, execute, get, lcd, local, put
 from fabric.contrib.console import confirm
 import datetime
 import os
@@ -15,17 +15,6 @@ env.use_ssh_config = False
 # fab will use the ssh keyfile referenced by the host alias, otherwise need to do what is
 # being done in dev to assign env a key_filename
 
-
-def development():
-    env.environment = 'dev'
-    env.hosts = '127.0.0.1'
-    env.port = 2222
-    env.user = 'vagrant'
-    env.host = 'openoversight-dev'
-    env.unprivileged_user = 'vagrant'
-    env.venv_dir = '/home/vagrant/oovirtenv'
-    env.code_dir = '/vagrant'
-    env.backup_dir = '/home/vagrant/openoversight_backup'
 
 
 def staging():
@@ -49,24 +38,30 @@ def production():
     env.backup_dir = '/home/nginx/openoversight_backup'
     env.s3bucket = 'openoversight-prod'
 
+env.roledefs = {
+    'staging': staging(),
+    'prod': 'openoversight.com',
+}
+
 
 def deploy():
-    with cd(env.code_dir):
-        run('su %s -c "git fetch && git status"' % env.unprivileged_user)
-        run('su %s -c "make cleanassets assets"' % env.unprivileged_user)
-        if confirm("Update to latest commit in this branch?", default=False):
-            run('su %s -c "git pull"' % env.unprivileged_user)
-            run('su %s -c "PATH=%s/bin:$PATH pip install -r requirements.txt"' % (env.unprivileged_user, env.venv_dir))
-            run('sudo systemctl restart openoversight')
+    with lcd(os.path.dirname(os.path.realpath(__file__))):
+        with cd(env.code_dir):
+            run('su %s -c "git fetch && git status"' % env.unprivileged_user)
+            execute(buildassets)
+            if confirm("Update to latest commit in this branch?", default=False):
+                run('su %s -c "git pull"' % env.unprivileged_user)
+                run('su %s -c "PATH=%s/bin:$PATH pip install -r requirements.txt"' % (env.unprivileged_user, env.venv_dir))
+                run('su %s -c "mkdir --parents %s/OpenOversight/app/static/dist"' % (env.unprivileged_user, env.code_dir))
+                put(local_path=os.path.join('OpenOversight', 'app', 'static', 'dist'),
+                    remote_path=os.path.join(env.code_dir, 'OpenOversight', 'app', 'static', 'dist')
+                    )
+                run('sudo systemctl restart openoversight')
 
 
 def migrate():
+    execute(deploy)
     with cd(env.code_dir):
-        run('su %s -c "git fetch && git status"' % env.unprivileged_user)
-        run('su %s -c "make cleanassets assets"' % env.unprivileged_user)
-        if confirm("Update to latest commit in this branch?", default=False):
-            run('su %s -c "git pull"' % env.unprivileged_user)
-            run('su %s -c "PATH=%s/bin:$PATH pip install -r requirements.txt"' % (env.unprivileged_user, env.venv_dir))
         if confirm("Apply any outstanding database migrations?", default=False):
             run('su %s -c "cd OpenOversight; FLASK_APP=OpenOversight.app %s/bin/flask db upgrade"' % (env.unprivileged_user, env.venv_dir))
             run('sudo systemctl restart openoversight')
@@ -83,3 +78,9 @@ def backup():
         get(remote_path="backup.tar.gz",
             local_path="./backup/backup-%s-%s.tar.gz"
                        % (env.environment, backup_datetime))
+
+
+def buildassets():
+    with lcd(os.path.dirname(os.path.realpath(__file__))):
+        local('mkdir -p OpenOversight/app/static/dist ; chmod go+rw OpenOversight/app/static/dist')
+        local('make cleanassets assets')

--- a/fabfile.py
+++ b/fabfile.py
@@ -16,7 +16,6 @@ env.use_ssh_config = False
 # being done in dev to assign env a key_filename
 
 
-
 def staging():
     env.environment = 'staging'
     env.user = 'root'
@@ -37,6 +36,7 @@ def production():
     env.code_dir = '/home/nginx/oovirtenv/OpenOversight'
     env.backup_dir = '/home/nginx/openoversight_backup'
     env.s3bucket = 'openoversight-prod'
+
 
 env.roledefs = {
     'staging': staging(),

